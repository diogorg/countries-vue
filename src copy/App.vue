<template>
  <div class="q-pa-md q-gutter-sm">
    <q-avatar icon="https://cdn.quasar.dev/img/avatar.png">
    </q-avatar>
  </div>
</template>

<script>

export default {
  name: 'App',

  components: {},

  setup() { }
}
</script>
