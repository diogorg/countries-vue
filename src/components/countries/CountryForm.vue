<template>
    <div class="filter-col q-pa-sm">
        <form v-on:keyup.enter="filterCountries" class="filter-form q-pa-sm q-pb-md shadow-1">
            <h3 class="q-pa-sm">Filter Countries</h3>
            <q-input outlined v-model="name" label="Country Name" class="q-ma-sm" />
            <q-select outlined v-model="region" :options="regions" label="Region" class="q-ma-sm" />
            <q-select class="q-ma-sm" outlined v-model="lang" use-input hide-selected fill-input input-debounce="0"
                :options="options" @filter="filterLangs" label="Languages" emit-value map-options>
                <template v-slot:no-option>
                    <q-item>
                        <q-item-section class="text-grey">
                            No results
                        </q-item-section>
                    </q-item>
                </template>
            </q-select>
            <q-btn @click="filterCountries" class="glossy q-ma-sm" color="deep-orange" icon="search" label="Filter" />
        </form>
    </div>
</template>

<script>
import { mapState } from 'vuex'
export default {
    name: 'CountryForm',
    data: () => ({
        name: '',
        region: null,
        lang: null,
        options: []
    }),
    mounted() {
        this.options = this.langs
    },
    methods: {
        filterLangs(val, update) {
            update(() => {
                if (!val) {
                    this.options = this.langs
                    return
                }
                const needle = val.toLowerCase()
                this.options = this.langs.filter(v => v.toLowerCase().indexOf(needle) > -1)
            })
        },
        async filterCountries() {
            this.$q.loading.show({
                message: 'Filtering the countries',
                boxClass: 'bg-grey-2 text-grey-9',
                spinnerColor: 'primary'
            })
            setTimeout(() => {
                this.$store.commit('filterCountries', { name: this.name, region: this.region, lang: this.lang })
                this.$q.loading.hide()
            }, 700)
        }
    },
    computed: {
        ...mapState(['regions', 'langs']),
    },
    components: {},
    setup() { }
}
</script>

<style scoped>
.filter-col {
    width: 300px;
    position: relative;
}

.filter-form {
    background: #fff;
    width: fit-content;
    min-height: 100px;
    border: 1px dashed #bbb;
    text-align: center;
    position: fixed;
}

.filter-form h3 {
    text-align: center;
    width: 100%;
    font-size: 21px;
    margin: 0px;
}
</style>