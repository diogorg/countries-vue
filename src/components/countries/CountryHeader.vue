<template>
    <q-toolbar class="bg-black text-white shadow-1 q-mb-sm">
        <q-toolbar-title>
            {{ user.name }} :: {{ user.app }}
        </q-toolbar-title>
        <div class="side-bar">
            <a target="_blank" :href="user.github" id="github-link">
                <font-awesome-icon icon="fa-brands fa-github" class="fa-2xl" />
            </a>
            <q-avatar>
                <img :src="user.avatar">
            </q-avatar>
        </div>
    </q-toolbar>
</template>

<script>
export default {
    name: 'CountryHeader',
    data() {
        return {
            user: {
                app: 'Countries Search',
                name: 'Diogo Gutierre',
                github: 'https://github.com/diogorg',
                avatar: 'https://avatars.githubusercontent.com/u/1072595?v=4'
            }
        }
    },
    components: {},

    setup() { }
}
</script>
<style scoped>
.side-bar {
    width: 100px;
    display: flex;
    justify-content: space-between;
    vertical-align: middle;
}

.side-bar>a {
    color: #fff;
    margin: auto 0px;
}
</style>
