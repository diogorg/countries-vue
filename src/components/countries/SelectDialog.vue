<template>
    <q-dialog v-model="dialog" v-if="selected != null">
        <q-card class="dialog-card">
            <q-card-section class="bg-blue-grey-2 text-black q-mb-none dialog-header">
                <div vertical class="text-h4">{{ selected.name.common }} <span v-if="selected.cioc">({{ selected.cioc
                }})</span></div>
                <div>
                    <img v-if="selected.coatOfArms.png" class="q-ma-sm" :src="selected.coatOfArms.png"
                        style="height: 32px; width: auto;" />
                    <img class="q-ma-sm" :src="selected.flags.png" style="height: 32px; width: auto;" />
                </div>
            </q-card-section>
            <div class="dialog-content">
                <div class="maps">
                    <iframe style="height: 100%;" width="260" id="gmap_canvas"
                        :src="`https://maps.google.com/maps?q=${selected.name.common}&ie=UTF8&iwloc=&output=embed`"
                        frameborder="0" scrolling="no" marginheight="0" marginwidth="0"></iframe>
                </div>
                <div class="infos q-pb-lg">
                    <p>
                        <span class="head">Official Name</span>
                        <span class="info">{{ selected.name.official }}</span>
                    </p>
                    <p>
                        <span class="head">Region</span>
                        <span class="info">{{ selected.region }} <span v-if="selected.subregion">- ({{
                                selected.subregion
                        }})</span></span>
                    </p>
                    <p v-if="selected.languages">
                        <span class="head">Language</span>
                        <span class="info">{{ showLangs(selected) }}</span>
                    </p>
                    <p v-if="selected.capital">
                        <span class="head">Capital</span>
                        <span class="info">{{ selected.capital[0] }}</span>
                    </p>
                    <p v-if="selected.currencies">
                        <span class="head">Currency</span>
                    <ul class="info q-ma-none" style="padding-left: 10px;">
                        <li v-for="(c, index) in selected.currencies" :key="index">{{ c.name }} ({{ c.symbol || '-' }})
                        </li>
                    </ul>
                    </p>
                    <p>
                        <span class="head">Total Area</span>
                        <span class="info">{{ selected.area.toLocaleString('en-US') }}</span>
                    </p>
                    <p v-if="selected.population">
                        <span class="head">Population</span>
                        <span class="info">{{ selected.population.toLocaleString('en-US') }}</span>
                    </p>
                    <p>
                        <span class="head">Borders</span>
                    <ul class="info q-ma-none borders-list" style="padding-left: 10px;">
                        <li v-for="(b, index) in selected.borders" :key="index">{{ b }}</li>
                        <li v-if="!selected.borders">No one</li>
                    </ul>
                    </p>
                    <p v-if="selected.maps.openStreetMaps">
                        <span class="head">Details</span>
                        <span class="info">
                            <q-btn size="11px" target="_blank" :href="selected.maps.openStreetMaps" align="between"
                                class="btn-fixed-width" color="primary" label="Open Detail" icon="flight_takeoff" />
                        </span>
                    </p>
                </div>
            </div>

            <q-card-actions align="right" class="close-button">
                <q-btn flat label="close" color="primary" @click="closeDialog" />
            </q-card-actions>
        </q-card>
    </q-dialog>
</template>

<script>
import { mapState, mapMutations } from 'vuex'
export default {
    name: 'SelectDialog',
    components: {},
    methods: {
        ...mapMutations(['closeDialog']),

        showLangs(country) {
            if (!country.languages) return '-'
            return Object.values(country.languages).join(' - ')
        },
    },
    computed: {
        ...mapState(['dialog', 'selected']),
    },
    setup() { }
}
</script>
<style lang="scss" scoped>
.dialog-card {
    max-height: max-content;
    min-width: 750px;
    min-height: 300px;
    overflow-y: hidden;
}

.dialog-card .dialog-header {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: space-between;
}

.close-button {
    right: 0px;
    position: absolute;
    bottom: 0px;
}

.dialog-content {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: flex-start;
    width: fit-content;
    overflow-y: visible;
    align-items: stretch;
    justify-content: space-between;
}

.dialog-content>.maps {
    width: 260px;
    background-color: #666;
    margin-right: 15px;
}

.dialog-content>.infos {
    margin-top: 10px;
    display: flex;
    flex-direction: column;
    flex-wrap: nowrap;
    align-items: flex-start;
    align-items: stretch;
    justify-content: space-between;
    font-size: 13px;
}

.dialog-content>.infos p {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: flex-start;
    align-items: stretch;
    justify-content: flex-start;
}

.dialog-content>.infos .head {
    width: 100px;
    text-align: right;
    margin-right: 20px;
    font-weight: bold;
    color: $cyan-8;
}

.dialog-content>.infos .info {
    text-align: left;
}

.dialog-card .dialog-header .text-h4 {
    line-height: auto;
    vertical-align: middle;
}

.borders-list li {
    display: inline-flex;
    white-space: nowrap;
    margin: 2px 4px;
}
</style>
