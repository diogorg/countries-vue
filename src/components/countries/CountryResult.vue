<template>
    <SelectDialog />
    <div class="q-pa-sm countries" style="">
        <q-card class="my-card q-ma-sm" v-for="(c, index) in countries" :key="index">
            <q-img :src="c.flags.png" class="flag">
                <div class="text-h6">{{ c.name.common }}</div>
            </q-img>

            <q-card-actions vertical align="center">
                <q-btn icon-right="open_in_new" label="Open" flat @click="selectCountry(c)" />
            </q-card-actions>
        </q-card>
        <div class="empty" v-if="countries.length == 0">Nothing found</div>
        <div class="total q-pr-xl" v-if="countries.length > 0">
            <q-chip color="deep-orange" text-color="white" icon="filter_alt">
                Total: <b class="q-pa-sm">{{ countries.length }}</b>
            </q-chip>
        </div>
    </div>
</template>

<script>
import { mapState } from 'vuex'
import SelectDialog from './SelectDialog.vue'
export default {
    name: 'CountryResult',
    components: {
        SelectDialog
    },
    methods: {
        selectCountry(country) {
            this.$store.commit('selectCountry', { country })
        },
    },
    computed: {
        ...mapState(['countries']),
    },
}
</script>

<style lang="scss" scoped>
.countries {
    min-height: 100px;
    display: flex;
    align-items: flex-start;
    flex-direction: row;
    flex-wrap: wrap;
    width: fit-content;
    overflow: auto;
    min-width: calc(100% - 300px);
    max-width: calc(100% - 300px);
    height: max-content;
}

.countries .flag {
    display: flex;
    max-height: 90px;
}

.my-card {
    width: 100%;
    max-width: 200px;
}

.empty {
    width: 100%;
    margin: 50px auto;
    text-align: center;
    color: $red-10;
    font-weight: bold;
    font-size: large;
}

.total {
    width: 100%;
    text-align: right;
}
</style>