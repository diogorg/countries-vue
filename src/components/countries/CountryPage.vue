<template>
    <CountryHeader />
    <div class="main q-pb-md">
        <CountryForm />
        <CountryResult />
    </div>
</template>

<script>
import CountryHeader from './CountryHeader.vue'
import CountryForm from './CountryForm.vue'
import CountryResult from './CountryResult.vue'
export default {
    name: 'CountryPage',
    components: {
        CountryHeader,
        CountryForm,
        CountryResult
    },
    data: () => ({}),
    beforeCreate() {
        const quasar = this.$q
        this.$store.commit('getAllCountriesAndRegions', { quasar })
    },
    setup() { }
}
</script>

<style scoped>
.main {
    height: 100%;
    width: 100%;
    display: flex;
    justify-content: space-between;
    vertical-align: top;
    flex-direction: row;
    align-items: stretch;
    overflow: auto;
    align-items: stretch;
}
</style>