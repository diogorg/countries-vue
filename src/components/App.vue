<template>
  <CountryPage />
</template>

<script>
import CountryPage from './countries/CountryPage.vue'
export default {
  name: 'App',

  components: {
    CountryPage
  },

  setup() { }
}
</script>
