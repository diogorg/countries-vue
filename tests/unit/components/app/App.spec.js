import { shallowMount } from '@vue/test-utils'
import App from '@/components/App.vue'
import { components } from '../countries/basic-components'

const global = {
    components,
}

describe('App.vue', () => {
    it('Should match snapshot', () => {
        const wrap = shallowMount(App, { global })
        expect(wrap.html()).toMatchSnapshot()
    })
})