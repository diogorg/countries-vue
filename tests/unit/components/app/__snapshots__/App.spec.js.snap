// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue Should match snapshot 1`] = `<country-page-stub></country-page-stub>`;
